File: frontend/src/services/LeaveService.ts
import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL;

// Interfaces cho DTOs
export interface LeaveRequestDto {
  maDon: number;
  maNv: number;
  tenNhanVien?: string;
  chucVu?: string;
  ngayNghi: string;
  lyDo?: string;
  trangThai?: string; // cho_duyet, da_duyet, tu_choi
}

export interface CreateLeaveRequest {
  maNv: number;
  ngayNghi: string;
  lyDo: string;
}

export interface ApproveLeaveRequest {
  trangThai: string; // da_duyet hoặc tu_choi
}

export interface PhantomReadDemoResult {
  lanDoc1: LeaveRequestDto[];
  lanDoc2: LeaveRequestDto[];
  message: string;
  startTime: string;
  endTime: string;
  hasPhantomRead: boolean;
}

class LeaveService {
  /**
   * Lấy token từ localStorage
   */
  private getToken(): string {
    return localStorage.getItem('access_token') || '';
  }

  /**
   * Tạo config với Authorization header
   */
  private getConfig() {
    return {
      headers: {
        Authorization: `Bearer ${this.getToken()}`,
      },
    };
  }


  async getAllLeaveRequests(): Promise<LeaveRequestDto[]> {
    try {
      // ========== MODE 1: NORMAL ==========
      const response = await axios.get<PhantomReadDemoResult>(
        `${API_BASE_URL}/leave-requests/phantom-demo`,
        this.getConfig()
      );
      return response.data.lanDoc2 || [];
      
      // ========== MODE 2: DEMO FIXED ==========
      // Quản lý dùng SERIALIZABLE lock. Khi quản lý đang đọc danh sách, nhân viên gửi đơn mới sẽ bị block

      // const response = await axios.get<PhantomReadDemoResult>(
      //   `${API_BASE_URL}/leave-requests/fixed-phantom-demo`,
      //   this.getConfig()
      // );
      // console.log('Fixed Phantom Read Demo Response:', response.data);
      // console.log('lanDoc2:', response.data.lanDoc2);
      // return response.data.lanDoc2 || [];

    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Lấy danh sách đơn nghỉ phép thất bại');
      }
      throw error;
    }
  }


  async getLeaveRequestById(maDon: number): Promise<LeaveRequestDto> {
    try {
      const response = await axios.get<LeaveRequestDto>(
        `${API_BASE_URL}/leave-requests/${maDon}`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Lấy thông tin đơn nghỉ phép thất bại');
      }
      throw error;
    }
  }


  async getLeaveRequestsByEmployee(maNv: number): Promise<LeaveRequestDto[]> {
    try {
      const response = await axios.get<LeaveRequestDto[]>(
        `${API_BASE_URL}/leave-requests/employee/${maNv}`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Lấy danh sách đơn nghỉ phép thất bại');
      }
      throw error;
    }
  }


  async getLeaveRequestsByStatus(trangThai: string): Promise<LeaveRequestDto[]> {
    try {
      const response = await axios.get<LeaveRequestDto[]>(
        `${API_BASE_URL}/leave-requests/status/${trangThai}`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Lấy danh sách đơn nghỉ phép thất bại');
      }
      throw error;
    }
  }


  async createLeaveRequest(request: CreateLeaveRequest): Promise<{ message: string; maDon: number }> {
    try {

      const response = await axios.post<{ message: string; maDon: number }>(
        `${API_BASE_URL}/leave-requests/will-block`,
        request,
        this.getConfig()
      );
      console.log(' Nhân viên gửi đơn thành công!');
      return response.data;

    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Tạo đơn nghỉ phép thất bại');
      }
      throw error;
    }
  }

  /**
   * Duyệt hoặc từ chối đơn nghỉ phép
   * @param maDon Mã đơn nghỉ phép
   * @param request Trạng thái mới (da_duyet hoặc tu_choi)
   */
  async approveLeaveRequest(maDon: number, request: ApproveLeaveRequest): Promise<{ message: string }> {
    try {
      const response = await axios.put<{ message: string }>(
        `${API_BASE_URL}/leave-requests/${maDon}/approve`,
        request,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Duyệt đơn nghỉ phép thất bại');
      }
      throw error;
    }
  }

  /**
   * Xóa đơn nghỉ phép
   * @param maDon Mã đơn nghỉ phép
   */
  async deleteLeaveRequest(maDon: number): Promise<{ message: string }> {
    try {
      const response = await axios.delete<{ message: string }>(
        `${API_BASE_URL}/leave-requests/${maDon}`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Xóa đơn nghỉ phép thất bại');
      }
      throw error;
    }
  }

  // =============================================
  // PHANTOM READ DEMO METHODS
  // =============================================

  /**
   * [DEMO] Đọc danh sách đơn nghỉ phép 2 lần (CÓ LỖI PHANTOM READ)
   */
  async getLeaveRequestsWithPhantomRead(): Promise<PhantomReadDemoResult> {
    try {
      const response = await axios.get<PhantomReadDemoResult>(
        `${API_BASE_URL}/leave-requests/phantom-demo`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Demo phantom read thất bại');
      }
      throw error;
    }
  }

  /**
   * [DEMO] Đọc danh sách đơn nghỉ phép 2 lần (ĐÃ FIX PHANTOM READ)
   */
  async getLeaveRequestsFixedPhantomRead(): Promise<PhantomReadDemoResult> {
    try {
      const response = await axios.get<PhantomReadDemoResult>(
        `${API_BASE_URL}/leave-requests/fixed-phantom-demo`,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Demo fixed phantom read thất bại');
      }
      throw error;
    }
  }

  /**
   * [DEMO] Tạo đơn nghỉ phép (NORMAL - không bị block)
   */
  async createLeaveRequestNormal(request: CreateLeaveRequest): Promise<{ message: string; maDon: number }> {
    try {
      const response = await axios.post<{ message: string; maDon: number }>(
        `${API_BASE_URL}/leave-requests/normal`,
        request,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Tạo đơn nghỉ phép (normal) thất bại');
      }
      throw error;
    }
  }

  /**
   * [DEMO] Tạo đơn nghỉ phép (SẼ BỊ BLOCK nếu có transaction đang lock)
   */
  async createLeaveRequestWillBlock(request: CreateLeaveRequest): Promise<{ message: string; maDon: number; waitedSeconds?: number }> {
    try {
      const response = await axios.post<{ message: string; maDon: number; waitedSeconds?: number }>(
        `${API_BASE_URL}/leave-requests/will-block`,
        request,
        this.getConfig()
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(error.response?.data?.message || 'Tạo đơn nghỉ phép (will block) thất bại');
      }
      throw error;
    }
  }
}

export default new LeaveService();

File: backend/Services/LeaveService.cs
using backend.Data;
using backend.DTOs;
using Microsoft.EntityFrameworkCore;
using System.Data;

namespace backend.Services;

/// <summary>
/// Service xử lý các nghiệp vụ liên quan đến Đơn nghỉ phép
/// </summary>
public class LeaveService : ILeaveService
{
    private readonly SportContext _context;
    private readonly ILogger<LeaveService> _logger;

    public LeaveService(SportContext context, ILogger<LeaveService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<List<LeaveRequestDto>> GetAllLeaveRequestsAsync()
    {
        try
        {
            var leaveRequests = await _context.DonNghiPheps
                .Include(d => d.MaNvNavigation)
                .ToListAsync();

            return leaveRequests
                .OrderByDescending(d => d.NgayNghi)
                .Select(d => new LeaveRequestDto
                {
                    MaDon = d.MaDon,
                    MaNv = d.MaNv ?? 0,
                    TenNhanVien = d.MaNvNavigation != null ? d.MaNvNavigation.HoTen : null,
                    ChucVu = d.MaNvNavigation != null ? d.MaNvNavigation.ChucVu : null,
                    NgayNghi = d.NgayNghi.HasValue ? d.NgayNghi.Value.ToDateTime(TimeOnly.MinValue) : default(DateTime),
                    LyDo = d.LyDo,
                    TrangThai = d.TrangThai
                })
                .ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error getting all leave requests: {ex.Message}");
            throw;
        }
    }

    public async Task<LeaveRequestDto?> GetLeaveRequestByIdAsync(int maDon)
    {
        try
        {
            var leaveRequest = await _context.DonNghiPheps
                .Include(d => d.MaNvNavigation)
                .Where(d => d.MaDon == maDon)
                .Select(d => new LeaveRequestDto
                {
                    MaDon = d.MaDon,
                    MaNv = d.MaNv ?? 0,
                    TenNhanVien = d.MaNvNavigation != null ? d.MaNvNavigation.HoTen : null,
                    ChucVu = d.MaNvNavigation != null ? d.MaNvNavigation.ChucVu : null,
                    NgayNghi = d.NgayNghi.HasValue ? d.NgayNghi.Value.ToDateTime(TimeOnly.MinValue) : default(DateTime),
                    LyDo = d.LyDo,
                    TrangThai = d.TrangThai
                })
                .FirstOrDefaultAsync();

            return leaveRequest;
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error getting leave request {maDon}: {ex.Message}");
            throw;
        }
    }

    public async Task<List<LeaveRequestDto>> GetLeaveRequestsByEmployeeAsync(int maNv)
    {
        try
        {
            // Fetch data first to avoid DateOnly translation issues
            var leaveRequests = await _context.DonNghiPheps
                .Include(d => d.MaNvNavigation)
                .Where(d => d.MaNv == maNv)
                .ToListAsync();

            // Sort and convert in memory
            return leaveRequests
                .OrderByDescending(d => d.NgayNghi)
                .Select(d => new LeaveRequestDto
                {
                    MaDon = d.MaDon,
                    MaNv = d.MaNv ?? 0,
                    TenNhanVien = d.MaNvNavigation != null ? d.MaNvNavigation.HoTen : null,
                    ChucVu = d.MaNvNavigation != null ? d.MaNvNavigation.ChucVu : null,
                    NgayNghi = d.NgayNghi.HasValue ? d.NgayNghi.Value.ToDateTime(TimeOnly.MinValue) : default(DateTime),
                    LyDo = d.LyDo,
                    TrangThai = d.TrangThai
                })
                .ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error getting leave requests for employee {maNv}: {ex.Message}");
            throw;
        }
    }

    public async Task<List<LeaveRequestDto>> GetLeaveRequestsByStatusAsync(string trangThai)
    {
        try
        {
            // Fetch data first to avoid DateOnly translation issues
            var leaveRequests = await _context.DonNghiPheps
                .Include(d => d.MaNvNavigation)
                .Where(d => d.TrangThai == trangThai)
                .ToListAsync();

            // Sort and convert in memory
            return leaveRequests
                .OrderByDescending(d => d.NgayNghi)
                .Select(d => new LeaveRequestDto
                {
                    MaDon = d.MaDon,
                    MaNv = d.MaNv ?? 0,
                    TenNhanVien = d.MaNvNavigation != null ? d.MaNvNavigation.HoTen : null,
                    ChucVu = d.MaNvNavigation != null ? d.MaNvNavigation.ChucVu : null,
                    NgayNghi = d.NgayNghi.HasValue ? d.NgayNghi.Value.ToDateTime(TimeOnly.MinValue) : default(DateTime),
                    LyDo = d.LyDo,
                    TrangThai = d.TrangThai
                })
                .ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error getting leave requests by status {trangThai}: {ex.Message}");
            throw;
        }
    }

    public async Task<int> CreateLeaveRequestAsync(CreateLeaveRequest request)
    {
        try
        {
            // Kiểm tra nhân viên có tồn tại không
            var employee = await _context.NhanViens.FindAsync(request.MaNv);
            if (employee == null)
            {
                throw new InvalidOperationException("Nhân viên không tồn tại");
            }

            // Kiểm tra ngày nghỉ phải từ hiện tại trở đi
            if (request.NgayNghi.Date < DateTime.Now.Date)
            {
                throw new InvalidOperationException("Ngày nghỉ phải từ hôm nay trở đi");
            }

            var leaveRequest = new DonNghiPhep
            {
                MaNv = request.MaNv,
                NgayNghi = DateOnly.FromDateTime(request.NgayNghi),
                LyDo = request.LyDo,
                TrangThai = "cho_duyet" // Mặc định là chờ duyệt
            };

            _context.DonNghiPheps.Add(leaveRequest);
            await _context.SaveChangesAsync();

            _logger.LogInformation($"Leave request created successfully with ID: {leaveRequest.MaDon}");
            return leaveRequest.MaDon;
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error creating leave request: {ex.Message}");
            throw;
        }
    }

    public async Task<bool> ApproveLeaveRequestAsync(int maDon, ApproveLeaveRequest request)
    {
        try
        {
            var leaveRequest = await _context.DonNghiPheps.FindAsync(maDon);
            if (leaveRequest == null)
            {
                throw new InvalidOperationException("Đơn nghỉ phép không tồn tại");
            }

            // Kiểm tra trạng thái hiện tại
            if (leaveRequest.TrangThai != "cho_duyet")
            {
                throw new InvalidOperationException("Đơn nghỉ phép đã được xử lý");
            }

            leaveRequest.TrangThai = request.TrangThai;
            await _context.SaveChangesAsync();

            _logger.LogInformation($"Leave request {maDon} updated to status: {request.TrangThai}");
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error approving leave request {maDon}: {ex.Message}");
            throw;
        }
    }

    public async Task<bool> DeleteLeaveRequestAsync(int maDon)
    {
        try
        {
            var leaveRequest = await _context.DonNghiPheps.FindAsync(maDon);
            if (leaveRequest == null)
            {
                throw new InvalidOperationException("Đơn nghỉ phép không tồn tại");
            }

            // Chỉ cho phép xóa đơn chưa duyệt hoặc bị từ chối
            if (leaveRequest.TrangThai == "da_duyet")
            {
                throw new InvalidOperationException("Không thể xóa đơn nghỉ phép đã được duyệt");
            }

            _context.DonNghiPheps.Remove(leaveRequest);
            await _context.SaveChangesAsync();

            _logger.LogInformation($"Leave request {maDon} deleted successfully");
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error deleting leave request {maDon}: {ex.Message}");
            throw;
        }
    }

    // =============================================
    // PHANTOM READ DEMO METHODS
    // =============================================

    public async Task<PhantomReadDemoResult> GetLeaveRequestsWithPhantomReadAsync()
    {
        var result = new PhantomReadDemoResult
        {
            StartTime = DateTime.Now,
            Message = "Demo Phantom Read (CÓ LỖI) - Đọc 2 lần với delay 5 giây"
        };

        try
        {
            using (var connection = _context.Database.GetDbConnection())
            {
                await connection.OpenAsync();
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "sp_GetLeaveRequests_WithPhantomRead";
                    command.CommandType = System.Data.CommandType.StoredProcedure;
                    command.CommandTimeout = 30;

                    using (var reader = await command.ExecuteReaderAsync())
                    {
                        // Đọc kết quả lần 1
                        while (await reader.ReadAsync())
                        {
                            result.LanDoc1.Add(new LeaveRequestDto
                            {
                                MaDon = reader.GetInt32(0),
                                MaNv = reader.GetInt32(1),
                                TenNhanVien = reader.IsDBNull(2) ? null : reader.GetString(2),
                                ChucVu = reader.IsDBNull(3) ? null : reader.GetString(3),
                                NgayNghi = reader.GetDateTime(4),
                                LyDo = reader.IsDBNull(5) ? null : reader.GetString(5),
                                TrangThai = reader.IsDBNull(6) ? null : reader.GetString(6)
                            });
                        }

                        // Chuyển sang result set tiếp theo (lần đọc 2)
                        await reader.NextResultAsync();
                        while (await reader.ReadAsync())
                        {
                            result.LanDoc2.Add(new LeaveRequestDto
                            {
                                MaDon = reader.GetInt32(0),
                                MaNv = reader.GetInt32(1),
                                TenNhanVien = reader.IsDBNull(2) ? null : reader.GetString(2),
                                ChucVu = reader.IsDBNull(3) ? null : reader.GetString(3),
                                NgayNghi = reader.GetDateTime(4),
                                LyDo = reader.IsDBNull(5) ? null : reader.GetString(5),
                                TrangThai = reader.IsDBNull(6) ? null : reader.GetString(6)
                            });
                        }
                    }
                }
            }

            result.EndTime = DateTime.Now;
            result.HasPhantomRead = result.LanDoc2.Count > result.LanDoc1.Count;
            
            if (result.HasPhantomRead)
            {
                result.Message += $" | ⚠️ PHANTOM READ DETECTED: Lần 1 có {result.LanDoc1.Count} đơn, lần 2 có {result.LanDoc2.Count} đơn";
            }
            else
            {
                result.Message += $" | ✓ Không phát hiện phantom read (cả 2 lần đều có {result.LanDoc1.Count} đơn)";
            }

            return result;
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error in GetLeaveRequestsWithPhantomReadAsync: {ex.Message}");
            throw;
        }
    }

    public async Task<PhantomReadDemoResult> GetLeaveRequestsFixedPhantomReadAsync()
    {
        var result = new PhantomReadDemoResult
        {
            StartTime = DateTime.Now,
            Message = "Demo ĐÃ FIX Phantom Read - Sử dụng SERIALIZABLE isolation level"
        };

        try
        {
            using (var connection = _context.Database.GetDbConnection())
            {
                await connection.OpenAsync();
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "sp_GetLeaveRequests_FixedPhantomRead";
                    command.CommandType = System.Data.CommandType.StoredProcedure;
                    command.CommandTimeout = 30;

                    using (var reader = await command.ExecuteReaderAsync())
                    {
                        // Đọc kết quả lần 1
                        while (await reader.ReadAsync())
                        {
                            result.LanDoc1.Add(new LeaveRequestDto
                            {
                                MaDon = reader.GetInt32(0),
                                MaNv = reader.GetInt32(1),
                                TenNhanVien = reader.IsDBNull(2) ? null : reader.GetString(2),
                                ChucVu = reader.IsDBNull(3) ? null : reader.GetString(3),
                                NgayNghi = reader.GetDateTime(4),
                                LyDo = reader.IsDBNull(5) ? null : reader.GetString(5),
                                TrangThai = reader.IsDBNull(6) ? null : reader.GetString(6)
                            });
                        }

                        // Chuyển sang result set tiếp theo (lần đọc 2)
                        await reader.NextResultAsync();
                        while (await reader.ReadAsync())
                        {
                            result.LanDoc2.Add(new LeaveRequestDto
                            {
                                MaDon = reader.GetInt32(0),
                                MaNv = reader.GetInt32(1),
                                TenNhanVien = reader.IsDBNull(2) ? null : reader.GetString(2),
                                ChucVu = reader.IsDBNull(3) ? null : reader.GetString(3),
                                NgayNghi = reader.GetDateTime(4),
                                LyDo = reader.IsDBNull(5) ? null : reader.GetString(5),
                                TrangThai = reader.IsDBNull(6) ? null : reader.GetString(6)
                            });
                        }
                    }
                }
            }

            result.EndTime = DateTime.Now;
            result.HasPhantomRead = result.LanDoc2.Count > result.LanDoc1.Count;
            
            result.Message += $" | ✓ ĐÃ FIX: Lần 1 có {result.LanDoc1.Count} đơn, lần 2 có {result.LanDoc2.Count} đơn (không thay đổi)";

            return result;
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error in GetLeaveRequestsFixedPhantomReadAsync: {ex.Message}");
            throw;
        }
    }

    public async Task<int> CreateLeaveRequestNormalAsync(CreateLeaveRequest request)
    {
        try
        {
            using (var connection = _context.Database.GetDbConnection())
            {
                await connection.OpenAsync();
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "sp_CreateLeaveRequest_Normal";
                    command.CommandType = System.Data.CommandType.StoredProcedure;

                    // Add parameters
                    var paramMaNv = command.CreateParameter();
                    paramMaNv.ParameterName = "@MaNv";
                    paramMaNv.Value = request.MaNv;
                    command.Parameters.Add(paramMaNv);

                    var paramNgayNghi = command.CreateParameter();
                    paramNgayNghi.ParameterName = "@NgayNghi";
                    paramNgayNghi.Value = request.NgayNghi.Date;
                    command.Parameters.Add(paramNgayNghi);

                    var paramLyDo = command.CreateParameter();
                    paramLyDo.ParameterName = "@LyDo";
                    paramLyDo.Value = request.LyDo;
                    command.Parameters.Add(paramLyDo);

                    var paramOutput = command.CreateParameter();
                    paramOutput.ParameterName = "@MaDonOutput";
                    paramOutput.Direction = System.Data.ParameterDirection.Output;
                    paramOutput.DbType = System.Data.DbType.Int32;
                    command.Parameters.Add(paramOutput);

                    await command.ExecuteNonQueryAsync();

                    return (int)paramOutput.Value;
                }
            }
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error in CreateLeaveRequestNormalAsync: {ex.Message}");
            throw;
        }
    }

    public async Task<int> CreateLeaveRequestWillBeBlockedAsync(CreateLeaveRequest request)
    {
        try
        {
            using (var connection = _context.Database.GetDbConnection())
            {
                await connection.OpenAsync();
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "sp_CreateLeaveRequest_WillBeBlocked";
                    command.CommandType = System.Data.CommandType.StoredProcedure;
                    command.CommandTimeout = 30; // Timeout 30 giây để chờ lock

                    // Add parameters
                    var paramMaNv = command.CreateParameter();
                    paramMaNv.ParameterName = "@MaNv";
                    paramMaNv.Value = request.MaNv;
                    command.Parameters.Add(paramMaNv);

                    var paramNgayNghi = command.CreateParameter();
                    paramNgayNghi.ParameterName = "@NgayNghi";
                    paramNgayNghi.Value = request.NgayNghi.Date;
                    command.Parameters.Add(paramNgayNghi);

                    var paramLyDo = command.CreateParameter();
                    paramLyDo.ParameterName = "@LyDo";
                    paramLyDo.Value = request.LyDo;
                    command.Parameters.Add(paramLyDo);

                    var paramOutput = command.CreateParameter();
                    paramOutput.ParameterName = "@MaDonOutput";
                    paramOutput.Direction = System.Data.ParameterDirection.Output;
                    paramOutput.DbType = System.Data.DbType.Int32;
                    command.Parameters.Add(paramOutput);

                    await command.ExecuteNonQueryAsync();

                    return (int)paramOutput.Value;
                }
            }
        }
        catch (Exception ex)
        {
            _logger.LogError($"Error in CreateLeaveRequestWillBeBlockedAsync: {ex.Message}");
            throw;
        }
    }
}
